Integrate jpo-utils Mongo Setup into CV Manager via Submodule

dmccoystephenson opened the pull request:
# PR Details

## Description
### Problem
The intersection-api Compose file defines its own MongoDB service, but a MongoDB service is already provided by the jpo-utils repository. Maintaining both is redundant.

### Solution
The redundant MongoDB service has been removed from intersection-api, and configurations have been updated to point to the MongoDB instance managed by jpo-utils, now targeting CV/Admin instead of ConflictMonitor. Unnecessary environment variables were also cleaned up, and related MongoDB resources were removed from the jpo-cvmanager repository.

## Testing
- Verified that the intersection_api service connects to MongoDB successfully.
- Verified that the connect service also connects to MongoDB successfully.

## Types of changes
- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)
- [x]  Local deployment change

## Checklist:
- [x] My changes require new environment variables:
  - [ ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [ ] My changes require updates to the documentation:
  - [ ] I have updated the documentation accordingly.
- [ ] My changes require updates and/or additions to the unit tests:
  - [ ] I have modified/added tests to cover my changes.
- [ ] All existing tests pass.

dmccoystephenson committed:
Use shared MongoDB configuration in docker-compose files

Replaced local MongoDB container definition with shared MongoDB configuration from `docker-compose-mongo.yml`. Updated dependencies and profiles to align with the new setup, simplifying and unifying service management.

dmccoystephenson committed:
Remove MongoDB initialization and management scripts

The MongoDB initialization, index management, and volume management scripts were removed, along with the related Dockerfile and keyfile.

dmccoystephenson committed:
Update MongoDB database and auth DB names in sample.env

Changed `CM_MONGO_AUTH_DB` and `CM_DATABASE_NAME` values from `ConflictMonitor` to `CV` in the environment file.

dmccoystephenson committed:
Removed 'mongodb_data_container'`' volume declaration in `docker-compose-intersection.yml`

dmccoystephenson committed:
Update MongoDB auth db and credentials in sample.env

Changed `CM_MONGO_AUTH_DB` to use the `admin` database. Updated default MongoDB usernames and passwords.

dmccoystephenson committed:
Update MongoDB database and authentication settings

Changed the MongoDB database name from 'ConflictMonitor' to 'CV' and the authentication database from 'ConflictMonitor' to 'admin' in the application.yaml file.

dmccoystephenson committed:
Set MongoDB database name in environment variables

Added the `SPRING_DATA_MONGODB_DATABASE` variable to the docker-compose file to configure the database name. This ensures proper connection setup for Spring Data MongoDB within the application.

dmccoystephenson committed:
Update MongoDB database and auth config in test YAML

Renamed the MongoDB database from `ConflictMonitor` to `CV` and updated the authentication database to `admin`.

dmccoystephenson committed:
Update MongoDB auth database and configuration variables

Replaced hardcoded MongoDB auth database and database name with environment variables (`CM_MONGO_AUTH_DB` and `CM_DATABASE_NAME`) for better flexibility and configurability. Updated `connect_start.sh` and `docker-compose-conflictmonitor.yml` to reflect these changes.

dmccoystephenson committed:
Update MongoDB database config to use environment variable

Replaced hardcoded database name with an environment variable `${CM_DATABASE_NAME}` in both `application.yaml` and `application-test.yaml`.

dmccoystephenson committed:
Merge branch 'docker/reference-kafka-in-jpo-utils' into docker/reference-mongo-in-jpo-utils

dmccoystephenson committed:
Add mongo and mongo-setup details to README.md

Updated the documentation to include information about the `mongo` and `mongo-setup` services for the intersection API. Clarified that these services are provided by the jpo-utils repository and require the `mongo_full` profile to be enabled.

dmccoystephenson commented on sample.env:
```diff
@@ -212,16 +212,12 @@ POSTMARK_SERVER_TOKEN=
-CM_MONGO_USER_PASSWORD=user
-
-CM_DATABASE_NAME=ConflictMonitor
+CM_MONGO_CONNECTOR_USERNAME=ode
+CM_MONGO_CONNECTOR_PASSWORD=replace_me
+CM_MONGO_API_USERNAME=ode
+CM_MONGO_API_PASSWORD=replace_me
```
Note: I'm not sure if these are the best credentials to use here, but testing showed that they worked.

Michael7371 replied:
We should override the environmental variable set and used in `jpo-utils` rather than use these environmental variables. If we set the same variable in the top level .env file it will use that when creating a user instead of what is set in the jpo-utils docker compose file. Also make sure to also override the `MONGO_DB_KEYFILE_STRING` variable.

So instead of `CM_MONGO_CONNECTOR_USERNAME` use `MONGO_READ_WRITE_USER` and stuff like that. JPO utils uses a shared user for the connector and API so those can just be set to the same thing.

dmccoystephenson replied:
Thank you for the input, I've implemented your suggestions!

dmccoystephenson commented on services/intersection-api/api/src/main/resources/application.yaml:
```diff
@@ -49,12 +49,12 @@ conflict:
   data:
     mongodb:
-      database: ConflictMonitor
+      database: ${CM_DATABASE_NAME:CV}
```
Note: This being hard-coded was resulting in failure to communicate with mongo after switching over to using the service provided by `jpo-utils`

dmccoystephenson reviewed (commented):
Self-review

dmccoystephenson committed:
Update environment variable names for MongoDB credentials

Replaced MongoDB credential variables with `MONGO_READ_WRITE_USER` and `MONGO_READ_WRITE_PASS` for consistency and clarity. Updated all relevant scripts, configuration files, and environment references to reflect.

dmccoystephenson committed:
Add placeholder for MongoDB keyfile string in sample.env

Included instructions to generate a secure MongoDB keyfile string using OpenSSL.

dmccoystephenson committed:
Merge branch 'docker/reference-kafka-in-jpo-utils' into docker/reference-mongo-in-jpo-utils

dmccoystephenson committed:
Merge branch 'docker/reference-kafka-in-jpo-utils' into docker/reference-mongo-in-jpo-utils

dmccoystephenson committed:
Merge remote-tracking branch 'origin/develop' into docker/reference-mongo-in-jpo-utils

dmccoystephenson committed:
docker: remove MongoDB service from docker-compose-intersection.yml

Michael7371 commented on docker-compose-conflictmonitor.yml:
```diff
@@ -1,3 +1,6 @@
+include:
+  - ./jpo-utils/docker-compose-mongo.yml
```
The `connect` service should be removed in favor of jpo-utils' `kafka-connect` service. After this, we should also be able to remove the `connect_start.sh` script. This will require including `./jpo-utils/docker-compose-connect.yml`

dmccoystephenson replied:
The `connect` service has been removed and we are now using kafka-connect from `./jpo-utils/docker-compose-connect.yml`. The `connect_start.sh` and corresponding Dockerfile has also been removed.

I am seeing the following in the logs of the `kafka-connect-setup` service, though:

> kafka-connect-setup-1  | Error: no matches found
kafka-connect-setup-1  | Waiting 10 sec for Kafka Connect to be ready (Attempt: 6/5)

Have you seen this before?

Michael7371 replied:
That seems like the jikkou connector management container is failing to reach the kafka connector [REST API](https://docs.confluent.io/platform/current/connect/references/restapi.html#kconnect-cluster). Can you check that you can query the kafka connect instance?

dmccoystephenson replied:
After copying `sample.env` to `.env` for the jpo-utils submodule, this was no longer occurring.

Michael7371 commented on sample.env:
```diff
@@ -4,8 +4,8 @@
-# basic, webapp, kafka_full, intersection, intersection_no_api, conflictmonitor, addons, obu_ota
-COMPOSE_PROFILES=basic,webapp,kafka_full,intersection
+# basic, webapp, mongo_full, kafka_full, intersection, intersection_no_api, conflictmonitor, addons, obu_ota
+COMPOSE_PROFILES=basic,webapp,mongo_full,kafka_full,intersection
```
I would recommend including the `kafka_connect_standalone` profile after including the `docker-compose-connect` compose file.

dmccoystephenson replied:
This is now included!

Michael7371 commented on sample.env:
```diff
@@ -186,7 +186,7 @@ DB_HOST_PORT=27017
 KAFKA_BROKER_PORT=9092
 
-CM_MONGO_AUTH_DB=ConflictMonitor
+CM_MONGO_AUTH_DB=admin
```
I would almost recommend removing this variable as it won't propagate down into jpo-utils properly. The auth DB for MongoDB should always be the `admin` database.

dmccoystephenson replied:
The `intersection-api` service in docker-compose-intersection.yml references this environment variable (and so does the application.yaml files in the intersection-api resources). Should we just hardcode 'admin' as the value?

Michael7371 replied:
I'd be okay with removing it from the `sample.env` now that the fallback value in the intersection API for the auth DB is `admin`.

dmccoystephenson replied:
Removed!

Michael7371 reviewed (changes requested):
Looks like a good start. I think we should use the jpo-utils connect module instead of duplicating it in the cv-manager custom one.

dmccoystephenson committed:
docker: remove custom Kafka Connect service and related scripts

Eliminated the custom Kafka Connect service setup, Dockerfile, and scripts to simplify the project setup. Updated the docker-compose configuration to utilize a shared Kafka Connect setup. Adjusted the environment profiles to reflect the changes.

dmccoystephenson committed:
docker: update MongoDB authentication database reference in configuration files

Michael7371 reviewed (changes requested):
It looks like none of the sample data from the [resources/mongo_scripts](https://github.com/CDOT-CV/jpo-cvmanager/tree/docker/reference-mongo-in-jpo-utils/resources/mongo_scripts) folder isn't being loaded in anymore. We should probably either remove these scripts or alter the data loading process to load this data at runtime.

dmccoystephenson committed:
chore: remove mongo scripts in favor of jpo-utils equivalents

dmccoystephenson commented:
> It looks like none of the sample data from the [resources/mongo_scripts](https://github.com/CDOT-CV/jpo-cvmanager/tree/docker/reference-mongo-in-jpo-utils/resources/mongo_scripts) folder isn't being loaded in anymore. We should probably either remove these scripts or alter the data loading process to load this data at runtime.

Good call! I've removed the mongo scripts in favor of the jpo-utils equivalents.

Michael7371 commented:
> > It looks like none of the sample data from the [resources/mongo_scripts](https://github.com/CDOT-CV/jpo-cvmanager/tree/docker/reference-mongo-in-jpo-utils/resources/mongo_scripts) folder isn't being loaded in anymore. We should probably either remove these scripts or alter the data loading process to load this data at runtime.
> 
> Good call! I've removed the mongo scripts in favor of the jpo-utils equivalents.

Looks like this will remove the sample ProcessedBsm injection for testing the v2x viewer, as well as some SSM and SRM data. @drewjj do you think this matters?

drewjj commented on docker-compose-intersection.yml:
```diff
@@ -63,89 +64,9 @@ services:
-      - INTERSECTION_API_ENABLE_API=${INTERSECTION_API_ENABLE_API}
-      - INTERSECTION_API_ENABLE_EMAILER=${INTERSECTION_API_ENABLE_EMAILER}
-      - INTERSECTION_API_ENABLE_REPORTS=${INTERSECTION_API_ENABLE_REPORTS}
-      - INSERT_SAMPLE_DATA=${INSERT_SAMPLE_DATA}
```
We should remove this INSERT_SAMPLE_DATA variable from the sample.env file if it is no longer used.

dmccoystephenson replied:
Removed!

Michael7371 replied:
@dmccoystephenson can we add documentation on the `MONGO_SAMPLE_DATA_RELATIVE_PATH` variable in jpo-utils to instruct the user how to not startup with sample data? It seems like if that variable is set to nothing, then it will skip the import.

dmccoystephenson replied:
Added a comment to the `sample.env`

Michael7371 replied:
That looks great!

drewjj reviewed (approved):
Everything runs. The environment variables are a little tricky with all the different variations of this we have but I was able to get it working. I think removing the no longer used environment variables from the sample.env helps make things clearer.

drewjj commented:
> > > It looks like none of the sample data from the [resources/mongo_scripts](https://github.com/CDOT-CV/jpo-cvmanager/tree/docker/reference-mongo-in-jpo-utils/resources/mongo_scripts) folder isn't being loaded in anymore. We should probably either remove these scripts or alter the data loading process to load this data at runtime.
> > 
> > 
> > Good call! I've removed the mongo scripts in favor of the jpo-utils equivalents.
> 
> Looks like this will remove the sample ProcessedBsm injection for testing the v2x viewer, as well as some SSM and SRM data. @drewjj do you think this matters?

Not sure. It is nice to be able to demo the sample data. I don't think we need the SRM and SSM data but the BSM is nice.

dmccoystephenson committed:
docker: remove `INSERT_SAMPLE_DATA` variable from sample.env

The `INSERT_SAMPLE_DATA` variable was unused and is no longer required in the environment configuration. Its removal helps simplify the `.env` file and avoid confusion about deprecated settings.

jacob6838 committed:
Adding notes about JPO-Utils docker profiles

jacob6838 committed:
Merge branch 'develop' into docker/reference-mongo-in-jpo-utils

jacob6838 committed:
Moving mongodump data to CV database

jacob6838 committed:
Updating jpo-utils submodule reference

jacob6838 committed:
Increaseing mongo data retention and adding dump path

jacob6838 reviewed (approved):
I pushed a few changes to the documentation and the sample data dump data/env vars, then tested and ran everything locally - worked great!
I added the following environment variables to the sample.env:
```
MONGO_DATA_RETENTION_SECONDS=31536000 # 1 year
MONGO_DATABASE_MAX_TTL_RETENTION_SECONDS=31536000 # 1 year
MONGO_SAMPLE_DATA_RELATIVE_PATH=../resources/mongodumps/dump_2024_08_20
```

jacob6838 committed:
Re-adding MONGO_DATABASE_MAX_TTL_RETENTION_SECONDS override

jacob6838 committed:
Merge branch 'develop' into docker/reference-mongo-in-jpo-utils

Michael7371 reviewed (approved):
Looks good!

dmccoystephenson committed:
docs: add comment for MONGO_SAMPLE_DATA_RELATIVE_PATH env var
